core/app/models/spree/address.rb

```diff
@@ -157,11 +157,17 @@ def country_iso=(iso)
 
     private
       def state_validate
-        # Skip state validation without country (also required)
-        # or when disabled by preference
-        return if country.blank? || !Spree::Config[:address_requires_state]
-        return unless country.states_required
+        # Skip state validation without country
+        return if country.blank?
 
+        # If states are not required we still check if state belongs to country
+        # If it doesn't we just remove it without adding state errors
+        unless country.states_required || readonly?
```

Comment:
Are you familiar with the immutable address work? I'm not sure this will work as you expect it to be, [as all persisted addresses are immutable](https://github.com/solidusio/solidus/blob/master/core/app/models/spree/address.rb#L147). It feels to me that there's a better place for this logic, perhaps [here](https://github.com/solidusio/solidus/blob/master/core/app/models/spree/address.rb#L46). cc @gvaughn also who may want to take a look at this.